### PR TITLE
Reimplement has_access decorator.

### PR DESCRIPTION
**Pull request by bkyryliuk**
I moved has_access to utils and extended the redirect upon successful login to provide next arg that can be used by the login implementation to redirect user back to the original path.

I plan to contribute it back to the FAB later on.

Partially solves: https://github.com/airbnb/superset/issues/1303

It can be used then in the login implementation
```
from flask import request, url_for

def redirect_url():
    return request.args.get('next') or request.referrer or url_for('index')


class AuthRemoteUserView(AuthRemoteUserView):
    @expose('/login/')
    def login(self):
      .....
      if success:
         return redirect(self.redirect_url())
```

Additional links:
* http://flask.pocoo.org/docs/0.12/reqcontext/
*  http://flask.pocoo.org/docs/0.12/api/#flask.Request.path

Reviewers:
* @mistercrunch 
* @vera-liu 
* @ascott

### CODE REVIEW

File: superset/utils.py [hunk 1]
```diff
@@ -25,7 +25,13 @@
...
 from email.mime.application import MIMEApplication
 from email.utils import formatdate
-from flask import flash, Markup, render_template
+from flask import flash, Markup, render_template, url_for, redirect, request
+from flask_appbuilder.const import (
+    LOGMSG_ERR_SEC_ACCESS_DENIED,
+    FLAMSG_ERR_SEC_ACCESS_DENIED,
+    PERMISSION_PREFIX
+)
+from flask_appbuilder._compat import as_unicode
```
Comment (mistercrunch) [inline superset/utils.py]: why do we need this? seems like it could bite us in the future.

Comment (bkyryliuk) [inline superset/utils.py]: @mistercrunch - I just copied existing implementation:
https://github.com/dpgaspar/Flask-AppBuilder/blob/af095ac4094d068fa2dd467bd01fa8b95d697fc1/flask_appbuilder/security/decorators.py#L5

I plan to create another PR to path FAB is possible, there is a TODO for it.

**Comment by rpwils** [comment]
Has there been anymore progress on this issue?

**Comment by rpwils** [comment]
xrmx - This solution was not fully implemented in superset. The auth views in flask appbuilder do not take this change into account and they are not implemented in superset.  We have implemented oauth login without the user needing to click the login buttons. When this occurs there is a popup to warn the user that Access was denied which is implemented in the has_access method thus forcing us to over write both. Any thought on how to resolve this?

**Pull request merged**